video: improve the capture sample

josuah opened the pull request:
I am proposing a lot of changes, with the goal to let reviewers decide which ones are relevant, essential, and which one are un-necessary or unwanted, and then I can apply the decisions.

For now, only tested with `zephyr,video-sw-generator` on `native_sim`:

```
west build -b native_sim/native/64 -S video-sw-generator samples/drivers/video/capture/
west build -b native_sim/native/64 -S video-sw-generator tests/drivers/video/test_pattern/
```

Then `west flash` to see the results.

josuah committed:
samples: drivers: video: capture: use if() instead of #ifdef

When possible, make use of C if statements instead of preprocessor #ifdef.
This still allow conditional compilation by letting the compiler
garbage-collect unused code.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: use static initializers

When possible, use static initializer instead of memset() and
initialization at runtime. Fixes a bug where uninitialized memory of caps
would be used, in case drivers do not initialize them like they should.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: use app_ prefix

Use an "app_" prefix for all functions local to the application.
This avoids accidentally using "display_" or "video_" prefix that
are reserved for the respective area.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: use the same error handling idiom

Use the same error handling implementation as everywhere else in the
video area and the majority of Zephyr:

  ret = error();
  if (ret < 0) {
      return ret;
  }

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
tests: drivers: video: test_pattern: new test from the sample

Turn the test part of the sample code into a dedicated test into the
tests/ directory instead of just samples/.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: move code to functions

Move blocks of code into individual functions to make the sample more
modular, component-based.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: change log verbosity

Increase log verbosity to print one line per captured frames and make
it like in the README by default. This avoids giving the impression
that the capture sample does not work. Users desiring to get better
performance can disable logging or reduce verbosity from INF to WRN.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: support test pattern CID

Add an always-on CID for enabling the test pattern, which makes
it possible to use it in tests enabling the test first enabling
the pattern.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented on tests/drivers/video/test_pattern/src/main.c:
```diff
@@ -0,0 +1,265 @@
+	zassert(bsize <= CONFIG_VIDEO_BUFFER_POOL_SZ_MAX,
+		"buffers too large");
+
+	for (int i = 0; i < CONFIG_VIDEO_BUFFER_POOL_NUM_MAX; i++) {
```
More buffers than minimum gives elasticity to the pipeline which might in some cases be needed for things to work.

When multiple devices get involved, using `min_vbuf_count` can be needed, but might be more interesting to handle this in a different PR.

josuah commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -229,12 +231,29 @@ int main(void)
 
+	ret = video_get_frmival(video_dev, &frmival);
+	if (ret < 0 && ret != -ENOTSUP) {
+		LOG_ERR("Error while setting the frame interval");
```
Done

josuah commented on tests/drivers/video/test_pattern/src/main.c:
```diff
@@ -0,0 +1,266 @@
+
+ZTEST(test_pattern, test_pattern_10_frames)
+{
+	struct video_buffer *vbuf = &(struct video_buffer){
+		.type = VIDEO_BUF_TYPE_OUTPUT
+	};
```
We need `&&vbuf`.

josuah commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -134,7 +133,7 @@ int main(void)
 
 	LOG_INF("- Capabilities:");
-	while (caps.format_caps[i].pixelformat) {
+	for (int i = 0; caps.format_caps[i].pixelformat; i++) {
```
Done

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -151,55 +151,56 @@ int main(void)
-	fmt.height = CONFIG_VIDEO_FRAME_HEIGHT;
-#endif
+	if (CONFIG_VIDEO_SOURCE_CROP_WIDTH > 0) {
+		struct video_selection sel = {
```
Seems this sel is shadowing another struct video_selection sel at the beginning of the main function, which is making it a bit confusing

avolmat-st replied:
Moreover it imposes to set a CROP_WIDTH / CROP_HEIGHT in order to do compose, which is not the case before and not expected.

avolmat-st replied:
Is it on purpose to only check CONFIG_VIDEO_SOURCE_CROP_WIDTH > 0 ? and not HEIGHT as well ?

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -151,55 +151,56 @@ int main(void)
-#endif
 
-	if (strcmp(CONFIG_VIDEO_PIXEL_FORMAT, "")) {
+	if (CONFIG_VIDEO_FRAME_HEIGHT > 0) {
```
Behavior would change I think like that since fmt.width / fmt.height used at line 183 should be normally updated with request given by CONFIG_VIDEO_FRAME_WIDTH / HEIGHT.

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -128,7 +127,8 @@ int main(void)
+	ret = video_get_caps(video_dev, &caps);
+	if (ret < 0) {
 		LOG_ERR("Unable to retrieve video capabilities");
 		return 0;
```
In all error cases like this, should we make it clear that the sample run has been aborted following this error ?
It might not be that obvious that nothing else is ongoing after getting those error messages.
In case of display sample for example, it is clearly mentioned "Aborting sample."

josuah commented:
Sorry @CkovMk I also missed your message.

> As the zephyr video subsystem evolves, we may build more and more complicated pipelines. Do you think it's possible to make this sample more flexible?

Yes absolutely, not only the sample, page 27 of this slide, libMP is introduced:
https://static.sched.com/hosted_files/osseu2025/7d/Video4Zephyr_EOSS_2025_Final.pdf#page=27

And here is a video about it:
https://www.youtube.com/watch?v=bjqUerc5jJw

But in the meantime, there is also more complex hardware coming with video encoder and internal ISP.

But for an external ISP, I suspect this would be only devicetree-level interconnection of devices? With a single input stream, arriving onto the video capture chip, the application would not be having more devices involved.

With the current APIs, it is possible to have the application configure the format of the ISP, and the ISP configure the format of the sensor.

Though, if it is arbitrary application/user choice for both, agreed, this needs to have both devices configured from `main.c` or similar.

josuah committed:
samples: drivers: video: make capture sample failure more visible

Centralize exit of all helper functions and display a common message
that indicates clearly that the sample is not blocking on any function
but instead stopped running.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Here https://github.com/zephyrproject-rtos/zephyr/pull/97425#discussion_r2428784181 is one possible way to handle such kind of workflows

```
                          |   optional   |
                          |              |
<sensor> -> <receiver> -> <test transform> -> appli
                          |              |
```

```c
video_dst = video_dev;
prev = &video_dst;

#if DT_HAS_CHOSEN(zephyr_videoisp)
videoisp_src = *prev;
*prev = videoisp_dev;
prev = &videoisp_dst;
#endif

#if DT_HAS_CHOSEN(zephyr_videoscaler)
videoscaler_src = *prev;
*prev = videoscaler_dev;
prev = &videoscaler_dst;
#endif

#if DT_HAS_CHOSEN(zephyr_videoenc)
videoenc_src = *prev;
*prev = videoenc_dev;
prev = &videoenc_dst;
#endif

videousb_src = *prev;
*prev = videousb_dev;
```

But there is only 2 devices in the current PR so not adding it yet if that makes sense.

ngphibang commented on tests/drivers/video/test_pattern/testcase.yaml:
```diff
@@ -0,0 +1,15 @@
+      - drivers
+      - video
+    platform_allow:
+      - native_sim/native/64
```
Could we add mimxrt1170-EVK board ? This test is mainly made on this board where I collected the threshold values.

ngphibang reviewed (commented):
Regarding the commit "tests: drivers: video: test_pattern: new test from the sample", will the "pattern test" be automatically triggered by CI from now on or it's just moving the test to a new place ?

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -454,6 +455,13 @@ static DEVICE_API(video, video_sw_generator_driver_api) = {
+	int ret;
+
+	ret = video_init_ctrl(&data->ctrls.test_pattern, dev, VIDEO_CID_TEST_PATTERN,
+			      (struct video_ctrl_range){.min = 1, .max = 1, .step = 1, .def = 1});
```
For now I just initialized it to `video_init_menu_ctrl()`, and taking no effect when `VIDEO_CID_TEST_PATTERN` is set to "Disabled".

The purpose was just to allow the test to call the `VIDEO_CID_TEST_PATTERN` of the video source, so that it is possible to test video-sw-generator the same way as the sensors...

Instead, should I not expose `VIDEO_CID_TEST_PATTERN` at all, and make the test ignore the failure when trying to test the test pattern?

ngphibang commented on tests/drivers/video/test_pattern/src/main.c:
```diff
@@ -0,0 +1,258 @@
+		ret = video_enqueue(video_dev, vbuf);
+		zassert_ok(ret);
+	}
+}
```
In real tests, sometimes I see the very first frames does not have the color bar falling into the defined threshold, maybe due to the sensor instability. In the original harness test, it only requires having one frame passing the test. So, I think this should be relaxed. Should we increase the loop iteration and break when see one (or some) frames that is OK and consider that the test has passed ?

josuah replied:
I updated the sample to have a Kconfig for the number of frames to test in total an the number of frames that need to be valid for the test to pass (threshold).

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -93,17 +93,21 @@ int main(void)
 	};
 #endif
+	struct video_caps caps = {
+		.type = VIDEO_BUF_TYPE_OUTPUT,
```
I understand the purpose of doing this assignment here so that more drivers would work. But, in a sense isn't this "hidding" potential failure of drivers ? Maybe doing so is ok, to have this sample do all it can to make the sample work, but, we should also start thinking about adding a new test / compliance video app which check the behaviors of drivers we are in charge of in this subsystem. Some thing a bit like v4l2-compliance, which does variable calls on the API to see if the drivers respond as expected.

josuah replied:
WRT `caps.type`, the assignment is already done in the current sample, here it just moves it to the static initializer:

https://github.com/zephyrproject-rtos/zephyr/blob/0876b5a2ded100ccc50af797c24f09eae70e4556/samples/drivers/video/capture/src/main.c#L100

https://github.com/zephyrproject-rtos/zephyr/blob/0876b5a2ded100ccc50af797c24f09eae70e4556/samples/drivers/video/capture/src/main.c#L130

Maybe ok to only change the code layout on this PR?

It is a good opportunity to review the semantics though.

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -132,87 +175,54 @@ int main(void)
 	}
 
+	/* Adjust video format according to the configuration */
 	if (CONFIG_VIDEO_FRAME_HEIGHT > 0) {
```
Shouldn't those formats update (based on Kconfig) be done in the main part rather than in this query format function ?
The query function seems to be more like gathering the current status. If the fmt structure is updated within this function this make it think that the current format is like this, which is not correct since only the structure has been changed and this hasn't been set to the driver.

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -265,51 +292,109 @@ int main(void)
-		vbuf = video_buffer_aligned_alloc(fmt.size, CONFIG_VIDEO_BUFFER_POOL_ALIGN,
-							K_NO_WAIT);
+		vbuf = video_buffer_aligned_alloc(fmt->size, CONFIG_VIDEO_BUFFER_POOL_ALIGN,
+						  K_FOREVER);
```
I think it is changed to K_NO_WAIT in recent code.

josuah committed:
samples: drivers: video: capture: use DEVICE_DT_GET_OR_NULL for display

Allow the display to also use if() instead of #if by leveraging the
DEVICE_DT_GET_OR_NULL() that permit display_dev to always be defined.
The compiler will const-fold all the unused variables and functions.

Signed-off-by: Josuah Demangeon <me@josuah.net>

avolmat-st reviewed (dismissed):
LGTM, thanks @josuah !!!

josuah commented:
@hakehuang 

You might be interested in seeing the new `tests/drivers/video/test_pattern` which is now separated from `samples/drivers/video/capture`.

I hope this suits you well.

josuah commented on tests/drivers/video/test_pattern/Kconfig:
```diff
@@ -0,0 +1,59 @@
+	int "Value used for the test pattern menu control"
+	default 1
+	help
+	  Some hardware support different typtes of test patteren in different order in the menu.
```
Good catch >_<
And agreed on better wording.

ngphibang reviewed (approved):
LGTM. Thanks @josuah !
